File: string/leetcode_20.java
package string;

import java.util.Stack;

public class leetcode_20 {
    public static void main(String[] args) {
        leetcode_20 lc = new leetcode_20();

        System.out.println(lc.isValid("()()"));
    }

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(')
                stack.push(')');
            else if (c == '{')
                stack.push('}');
            else if (c == '[')
                stack.push(']');
            else if (stack.isEmpty() || stack.pop() != c)
                return false;
        }
        return stack.isEmpty(); 
    }
}

File: string/LeetCode_20.kt
package string

import java.util.Stack

class LeetCode_20 {

    fun isValid(s: String): Boolean {
        if (s.isEmpty() || s.length % 2 != 0) {
            return false
        }

        val stack = Stack<Char>()

        s.forEach {
            when (it) {
                '(', '[', '{' -> stack.push(it)
                ']', '}', ')' -> {
                    if (stack.isEmpty() || (stack.peek() != it - 1 && stack.peek() != it - 2)) {
                        return false
                    }
                    stack.pop()
                }
            }
        }

        return stack.isEmpty()
    }
}

fun main() {
    // Input
    val s = "()"
    // Output
    LeetCode_20().isValid(s).let {
        println(it)
        require(it)
    }
}
